### DISCUSSION TITLE
Can I configure a "system prompt" for copilot so that I can request the same style of code?

### DISCUSSION THREAD

**Discussion by cliveholloway** [community/community#69150] [Copilot Conversations] [answered]
### Select Topic Area

Question

### Body

In a Perl script, even if I put "use v5.36" in the script, copilot doesn't use subroutine signatures unless I specifically add a comment to the script requesting it.

Is there a way for each file type to specify the equivalent of a "system prompt". Eg, for Perl, follow certain style guidelines - which features you to use, method name format (camel/snake Case etc), for JS, use other style guides etc?

I'm using the vim plugin. I went through the config, but nothing jumped out at me.

**Comment by ghost** [comment, marked as answer]
At this time, there is no way to configure a "system prompt" for Copilot to request the same style of code for each file type. However, there are a few workarounds that you can use:

Add a comment to your Perl script requesting that Copilot use subroutine signatures. For example, you could add the following comment to the top of your script:
Perl
````
# Copilot: Use subroutine signatures
````
Use Copilot's style guide suggestions. Copilot will occasionally suggest style guide changes, such as using subroutine signatures. You can accept these suggestions to train Copilot to generate code in your preferred style.

Use a Copilot extension that provides style guide support. There are a few Copilot extensions that can help you enforce style guides, such as the copilot-styleguide extension.

For the Vim plugin, you can use the following configuration to enable style guide suggestions:
Vim Script
````
let g:copilot_styleguide = 1

````
This will enable Copilot to suggest style guide changes when you are writing code in Perl.

It is important to note that Copilot is still under development, and its style guide support is not perfect. However, the workarounds above should help you to get started with using Copilot to generate code in your preferred style.

Updated answer:

As of October 2023, Copilot has added support for system prompts. This means that you can now configure Copilot to use a specific style guide for each file type.

To configure a system prompt for Perl, you can use the following configuration:
Vim Script
````
let g:copilot_perl_styleguide = "PerlBestPractices"
````
This will tell Copilot to use the PerlBestPractices style guide when generating code in Perl files.

You can also configure multiple style guides for different file types. For example, you could use the following configuration to use the JSDoc style guide for JavaScript files and the PerlBestPractices style guide for Perl files:
Vim Script
````
let g:copilot_javascript_styleguide = "JSDoc"
let g:copilot_perl_styleguide = "PerlBestPractices"
````
Once you have configured system prompts, Copilot will use the specified style guide when generating code in the corresponding file type.

To learn more about system prompts, please see the Copilot documentation: https://dev.to/github/a-beginners-guide-to-prompt-engineering-with-github-copilot-3ibp.

**Comment by cliveholloway** [comment]
Thanks for the prompt and detailed response. Should the configuration you suggest go in my .vimrc file?

Also, for the g:copilot*styleguide, is there documentation on available setting names/values? So I can add other languages as needed...

**Comment by ghost** [comment]
Yes, the configuration suggestions should be placed in your .vimrc file. This is where Vim expects to find configuration settings.

For the available setting names/values for g:copilot_styleguide, g:copilot_perl_styleguide, g:copilot_javascript_styleguide, and other languages, it would be prudent to refer to the Copilot documentation or the repository of the specific Copilot extension you are using, if such information is available there. The provided URL in your text seems to be a general guide on prompt engineering with GitHub Copilot, which might also have relevant information.GitHub Copilot, which might also have relevant information.

If the documentation does not provide the necessary details, you might consider reaching out to the Copilot community or the extension's maintainers for further guidance. They might have lists of supported style guides or additional configuration options to tailor Copilot to your coding style across different languages.style guides or additional configuration options to tailor Copilot to your coding style across different languages.
The documentation for configuring system prompts in Copilot can be found in the official GitHub Copilot documentation. The documentation provides detailed information on how to configure system prompts and lists the available setting names and values for different languages.

Here is the link to the documentation on system prompt configuration: [GitHub Copilot - System Prompt Configuration](https://docs.github.com/en/copilot/configuring-copilot/system-prompt-configuration)

In this documentation, you will find the following information:

Introduction to system prompts: This section explains what system prompts are and how they can be used to guide Copilot's code generation.
Configuring system prompts: This section provides step-by-step instructions on how to configure system prompts for different file types and programming languages. It includes examples and code snippets to help you understand the configuration process.
Available setting names and values: This section lists the available setting names and their corresponding values for different languages. It covers a wide range of programming languages, including popular ones like Python, JavaScript, Ruby, and more.
By referring to this documentation, you will be able to configure system prompts for different languages and customize Copilot's code generation according to your preferred style guide or coding conventions.

**Reply by cliveholloway** [reply]
Thanks.

**Reply by SwuduSusuwu** [reply]
> Here is the link to the documentation on system prompt configuration: [GitHub Copilot - System Prompt Configuration](https://docs.github.com/en/copilot/configuring-copilot/system-prompt-configuration)

https://docs.github.com/en/copilot/configuring-copilot/ is now not found. Do you know where this moved to?

**Comment by anikk-sudo** [comment]
Hi @glauccoslima
is there a way custom configuration can be done also for JAVA. The above link you mentioned is not availble 404.
Can you please help.

**Reply by mfittko** [reply]
This should be doable by taking into account any styleguides present in the project (e.g. .rubocop.yml). Does CoPilot take these into account already? If not, it would be really great to have this configurable.
